syntax = "proto2";

message Variable {
  required string name = 1;
  repeated float data_float32 = 2;
  repeated double data_float64 = 3;
  repeated int32 data_int32 = 4;
  repeated int64 data_int64 = 5;
  repeated int32 shape = 6;
  required string dtype = 7;
  optional RecursiveSequenceLength recursive_sequence_lengths = 8;
  message RecursiveSequenceLength {
    repeated SequenceLength sequence_lengths = 1;
    message SequenceLength {
      repeated int32 lengths = 1;
    }
  }
}

message UserInfo {
  required string token = 1;
  required string uid = 2;
}

message Features {
  required string token = 1;
  repeated Variable vars = 2;
  optional State state = 3;
};

message SaveInfo {
  required string token = 1;
  required string path = 2;
  repeated string feeded_var_names = 3;
  required string save_token = 4;
}

message NilRequest {
  required string token = 1;
  required State state = 2;
}

message NilResponse {
  required State state = 1;
}

message State {
  required bool succ = 1;
  optional string error_message = 2;
}

service FLExecutor {
  rpc execute_forward_host_part(UserInfo) returns (Features) {}
  rpc execute_backward_host_part(Features) returns (NilResponse) {}
  rpc save_persistables(SaveInfo) returns (NilResponse) {}
  rpc save_inference_model(SaveInfo) returns (NilResponse) {}
  rpc cancel_current_step(NilRequest) returns (NilResponse) {}
};
